internal/controllers/machine/machine_controller_inplace_update.go

```diff
@@ -0,0 +1,266 @@
+	return hasAnnotation
+}
+
+// isBootstrapConfigReadyForUpdate checks if the BootstrapConfig has the in-place update annotation.
+func (r *Reconciler) isBootstrapConfigReadyForUpdate(s *scope) bool {
+	if s.bootstrapConfig == nil {
+		return true
+	}
+	_, hasAnnotation := s.bootstrapConfig.GetAnnotations()[clusterv1.UpdateInProgressAnnotation]
+	return hasAnnotation
+}
+
+// callUpdateMachineHook calls the UpdateMachine runtime hook for the machine.
+func (r *Reconciler) callUpdateMachineHook(ctx context.Context, s *scope) (ctrl.Result, string, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Validate that exactly one extension is registered for the UpdateMachine hook.
+	// For the current iteration, we only support a single extension to ensure safe behavior.
+	// Support for multiple extensions will be introduced in a future iteration.
+	extensions, err := r.RuntimeClient.GetAllExtensions(ctx, runtimehooksv1.UpdateMachine, s.machine)
+	if err != nil {
+		return ctrl.Result{}, "", err
+	}
+
+	if len(extensions) == 0 {
+		return ctrl.Result{}, "", errors.New("no extensions registered for UpdateMachine hook")
+	}
+
+	if len(extensions) > 1 {
+		return ctrl.Result{}, "", errors.Errorf("multiple extensions registered for UpdateMachine hook: only one extension is supported in the current iteration, found %d extensions: %v", len(extensions), extensions)
+	}
+
+	request := &runtimehooksv1.UpdateMachineRequest{
+		Desired: runtimehooksv1.UpdateMachineRequestObjects{
+			Machine:               *cleanupMachine(s.machine),
+			InfrastructureMachine: runtime.RawExtension{Object: cleanupUnstructured(s.infraMachine)},
+		},
+	}
+
+	if s.bootstrapConfig != nil {
+		request.Desired.BootstrapConfig = runtime.RawExtension{Object: cleanupUnstructured(s.bootstrapConfig)}
+	}
+
+	response := &runtimehooksv1.UpdateMachineResponse{}
+
+	if err := r.RuntimeClient.CallAllExtensions(ctx, runtimehooksv1.UpdateMachine, s.machine, request, response); err != nil {
+		return ctrl.Result{}, "", err
+	}
+
+	if response.GetRetryAfterSeconds() != 0 {
+		log.Info(fmt.Sprintf("UpdateMachine hook requested retry after %d seconds", response.GetRetryAfterSeconds()))
+		return ctrl.Result{RequeueAfter: time.Duration(response.GetRetryAfterSeconds()) * time.Second}, response.GetMessage(), nil
+	}
+
+	log.Info("UpdateMachine hook completed successfully")
+	return ctrl.Result{}, response.GetMessage(), nil
+}
+
+// completeInPlaceUpdate removes in-place update annotations from InfraMachine, BootstrapConfig, Machine,
+// and then marks the UpdateMachine hook as done (removes it from pending-hooks annotation).
+func (r *Reconciler) completeInPlaceUpdate(ctx context.Context, s *scope) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	if err := r.removeInPlaceUpdateAnnotation(ctx, s.machine); err != nil {
+		return err
+	}
+
+	if s.infraMachine == nil {
+		return errors.New("InfraMachine must exist to complete in-place update")
+	}
```

Comment:
I'm wondering if, when we arrive at this stage (in-place completed), we should always try to completeInPlaceUpdate and just log a warning instead of failing when  infraMachine does not exists anymore

@alexander-demicev @sbueringer opinions?

Reply:
Since it's a cleanup after already finished in-place update, I'm up for logging a warning